### ISSUE TITLE
OculiX V5: Embedded Visual Intelligence pipeline for semantic UI understanding

### ISSUE THREAD

**Issue by julienmerconsulting** [oculix-org/Oculix#160]
## Problem

OculiX does visual testing via pixel matching: template matching, OCR, image comparison. It's deterministic but fragile — resolution change, theme switch, DPI shift, font change breaks scripts. No open-source visual testing tool embeds visual intelligence capable of **understanding** what it sees (component type, state, semantic context) without depending on an external LLM, cloud API, or paid tokens.

## Vision

Give OculiX an **embedded, local, open-source, LLM-free visual intelligence** capable of:

- Recognizing standard UI components (button, input, checkbox, dropdown, modal, tab, menu, icon, loader, slider, toggle...)
- Detecting component state (active, disabled, error, success, loading, selected)
- Classifying screen zones (header, navigation, sidebar, content, form, footer, dialog) — **rule-based from layout heuristics, no ML needed**
- Working cross-platform (Android, iOS, Web, Desktop Windows/macOS/Linux)
- Running locally without GPU, without cloud, without API calls

### Option A: Tiling pipeline (4 stages)

```
Screenshot → Tiling (64-128px) → OpenCV features → Conditional OCR → ONNX classification
```

- Tiles are autonomous analysis units with configurable overlap (25%)
- OpenCV extracts deterministic features per tile: contours (Canny), dominant color (HSV), texture (Sobel), content density, symmetry → ~15-20 dimension vector
- OCR runs ONLY on tiles with high text probability (~30% of tiles) — saves 70% compute
- ONNX model classifies: component type + state + confidence

**Limitation:** fixed-size tiles cut wide components (input fields, nav bars) into fragments classified independently.

### Option B: YOLO object detection (single pass) — to evaluate

```
Screenshot → YOLOv8-nano → bounding boxes with component type + confidence
```

- Detects components at arbitrary sizes and positions in one pass
- YOLOv8-nano: ~5ms per frame on CPU
- Eliminates tiling fragmentation problem
- Needs bounding-box annotated training data (not just tile labels)

**Recommendation:** evaluate both on Rico dataset, benchmark accuracy + speed, pick winner.

### Zone classification — rule-based (no ML)

Screen zones follow universal layout heuristics:
- Top 10% → header
- Bottom 8% → footer
- Left 15% (if distinct) → sidebar
- Center → main content
- Overlays with shadow/border → modal/dialog

No model needed. Saves ML budget for component detection.

### State detection — Phase 2 (separate from types)

Component state (active vs disabled vs error) is the hardest part:
- "Disabled" is often a subtle 10% opacity change
- "Hover" is **impossible** from a static screenshot (cursor-dependent transient state)
- "Focus" is a 2px outline

**Strategy:** ship component TYPE detection first (Phase 1). Add STATE detection as a separate training phase once types are stable. Remove "hover" from scope — not detectable from screenshots.

### Public sources (~500K+ base images)

| Dataset | Platform | Volume | License |
|---------|----------|--------|---------|
| Rico | Android mobile | 66K screenshots + hierarchies | CC BY 4.0 |
| Screen Annotation (Google) | Android mobile | 22K annotated | CC BY 4.0 |
| WebUI | Web | 400K rendered pages | Academic |
| ScreenSpot-Pro | Desktop (23 apps, 3 OS) | 1,581 screenshots HD | Open source |
| Desktop UI Element Dataset | Desktop | 100+ annotated screenshots | Open source |
| UI Components Detection (Roboflow) | Cross | 14 classes | Open source |
| AitW (Android in the Wild) | Android | Trajectories + screenshots | Google Research |

### ⚠ Cross-platform imbalance

Rico (66K mobile) vs ScreenSpot-Pro (1,581 desktop) = massive mobile bias. For a tool targeting POS/kiosks/desktop Windows, **desktop dataset enrichment is critical, not optional.** Community contributions and synthetic desktop screenshots must be prioritized.

### Data augmentation (UI-safe only)

**Valid:** multi-resolution resize, dark/light mode simulation, JPEG compression (50-90%), light Gaussian blur (VNC/remote), DPI scaling.

**Forbidden:** rotation, shear, perspective distortion, vertical flip, aggressive crop, salt-and-pepper noise. UIs don't rotate.

### Contribution format

```json
{
  "image_id": "contrib_001.png",
  "platform": "web", "os": "windows", "resolution": "1920x1080",
  "components": [
    { "type": "button", "state": "active", "zone": "header",
      "bbox": [120, 45, 200, 80], "text": "Submit", "confidence": 1.0 }
  ]
}
```

Hosted on Hugging Face: `oculix-org/visual-testing-dataset` (CC BY 4.0)

## Training

- **Infra:** Kaggle (free T4 GPU, 30h/week) — zero hardware investment
- **Stack:** PyTorch + torchvision + Albumentations + ONNX export
- **Model:** EfficientNet-B0 or MobileNetV3 fine-tuned, multi-task heads (type + state + confidence)
- **Iterations:** V1 on Rico only → V2 add WebUI + desktop → V3 community contributions
- **Target:** ONNX 20-50 MB, CPU inference <50ms per screen, <200 MB RAM
- **Model versioning:** downloaded separately from JAR (like Selenium browser drivers), independently updatable

### New module: `org.sikuli.script.visual`

- `VisualIntelligence.java` — pipeline orchestrator
- `TileAnalyzer.java` or `ObjectDetector.java` — depending on pipeline choice
- `ComponentClassifier.java` — ONNX model loading + inference
- `ScreenContext.java` — semantic screen representation

### New API

```java
// Find all active buttons
List<UIComponent> buttons = screen.findComponents(ComponentType.BUTTON, State.ACTIVE);

// Click the button containing "Submit"
screen.findComponent(ComponentType.BUTTON, "Submit").click();

// Assert a field is in error state
assert screen.findComponent(ComponentType.INPUT, "Email").getState() == State.ERROR;

// Full semantic analysis
ScreenContext ctx = screen.analyze();
ctx.getZone(Zone.HEADER).getComponents().forEach(System.out::println);
```

### Compatibility

- Existing pipeline (template matching, OCR) stays available
- Visual Intelligence is optional, enabled by config
- Auto-fallback: if model confidence < 0.5, falls back to classic pixel matching
- ONNX model downloaded separately, not bundled in main JAR

## Competitive positioning

| Criteria | Applitools | Testim | Percy | **OculiX V4** |
|----------|-----------|--------|-------|---------------|
| Cloud required | Yes | Yes | Yes | **No** |
| Token/API cost | Yes | Yes | Yes | **No** |
| Open source | No | No | No | **Yes** |
| Cross-platform | Partial | Web | Web | **Full** |
| Embedded intelligence | No (cloud) | No (cloud) | No (cloud) | **Yes (ONNX local)** |
| Works offline | No | No | No | **Yes** |
| LLM dependency | Variable | Yes | No | **No** |

## Roadmap

| Phase | Content | Estimate |
|-------|---------|----------|
| Phase 0 | Aggregate public datasets, unified format, publish HF | 2-3 weeks |
| Phase 1 | OpenCV pipeline (tiling + features) OR YOLO evaluation in Java | 3-4 weeks |
| Phase 2 | Train V1 component TYPE detection (Kaggle) | 3-4 weeks |
| Phase 3 | ONNX Runtime integration in OculiX, Java API | 2-3 weeks |
| Phase 4 | Add desktop datasets, train V2 + STATE detection | 3-4 weeks |
| Phase 5 | Community contribution tool, "annotate" mode in OculiX | 3 weeks |
| Phase 6 | Benchmarks, documentation, comms (article + HF) | 1 week |

**Realistic total: ~18-22 weeks (4.5-5.5 months)**

## Ecosystem connections

- **Hugging Face:** dataset + model on Hub, community visibility
- **Mistral / open-source models:** optional LLM call for high-level tasks, core visual testing works without
- **NVIDIA Nemotron Coalition:** trained model contributed as "visual testing" use case
- **Bruno Legeard / Smartesting:** concrete answer to testing AI agents without LLM dependency

---

*"The first visual testing tool with embedded visual intelligence — open source, local, LLM-free."*

**Comment by GeTOUO** [comment]
期待，我想需要类似于这样的能力：
```
UITree tree = screen.analyzeUITreeUntil(filterA, timeout);
tree.getPane(filterB).getButton(filterC).click();
```

**Comment by julienmerconsulting** [comment]
<div align="center">

[![Status](https://img.shields.io/badge/Status-API%20design%20acknowledged-1f883d?style=flat-square&logo=github)](.)
[![Module](https://img.shields.io/badge/Module-V4%20%C2%B7%20Visual%20Intelligence-8250df?style=flat-square&logo=opencv)](.)
[![Roadmap](https://img.shields.io/badge/Target-OculiX%20v4.0-0969da?style=flat-square&logo=apachemaven)](.)
[![Contributor](https://img.shields.io/badge/Contributor-%40GeTOUO-ff69b4?style=flat-square&logo=github)](.)
[![Sister%20issue](https://img.shields.io/badge/Sister%20issue-%23232%20%28Unicode%20type%29-1f883d?style=flat-square&logo=github)](../issues/232)
[![Ships%20with](https://img.shields.io/badge/Builds%20on-rc5%20%23233-8250df?style=flat-square&logo=github)](../issues/233)

</div>

你好 @GeTOUO — 谢谢你的提案 🙏

Your fluent API sketch is **exactly** the surface I want for V4:

```java
UITree tree = screen.analyzeUITreeUntil(filterA, timeout);
tree.getPane(filterB).getButton(filterC).click();
```

It captures three things the current pixel-matching API can't express today, and that I had only as fuzzy intuitions in #160:

| Capability | Today (V3) | Your proposal |
|---|---|---|
| **Wait until UI is in a known semantic state** | `wait(image, timeout)` — pixel-bound, brittle | `analyzeUITreeUntil(filter, timeout)` — semantic, theme-/DPI-/i18n-resilient |
| **Navigate the UI as a typed tree** | flat `find()` returning a Match (no parent/child) | `tree.getPane(...).getButton(...)` — typed, hierarchical, IDE-completion-friendly |
| **Filter by component category, not pixels** | filename of an image | `filter` predicate — by type, text content, state, zone, role |

This is the level of abstraction that turns OculiX from "image-recognition automation" into **semantic UI automation**. It's the exact gap between Sikulix-style scripts and what testers genuinely need in 2026.

## 🧩 How this slots into the V4 design (#160)

The V4 pipeline ends at `ScreenContext` — a flat list of detected components + zones. Your `UITree` is the missing **navigational view** on top of that detection result:

```
Screenshot
   ↓
Visual Intelligence pipeline (OpenCV / YOLO / ONNX)
   ↓
ScreenContext  — list of components with type, state, bbox, zone
   ↓ ✨ your contribution: the navigational tree on top
UITree         — typed parent/child navigation, fluent filters, await semantics
   ↓
script API     — getPane(...).getButton(...).click()
```

The tree structure follows naturally from the zone classifier (#160 — "rule-based zones, no ML needed"): `header → nav → button[]`, `content → form → input[]`, `dialog → ok-button`, etc. The `analyzeUITreeUntil` is just a fixed-point loop on the analyzer + a predicate.

## 📐 What the V4 API will likely look like (folding in your idea)

```java
// Wait until a visual condition is met (semantic, not pixel)
UITree tree = screen.analyzeUITreeUntil(
    t -> t.getDialog("Login") != null,
    Duration.ofSeconds(10)
);

// Hierarchical navigation, typed return values
tree.getPane(Zone.HEADER)
    .getButton(text("Submit"))
    .click();

// Mix of predicates: type + text + state
tree.findFirst(
    is(ComponentType.INPUT)
        .and(hasLabel("Email"))
        .and(inState(State.ERROR))
).getValue();
```

The `filter` is a `Predicate<Component>` so users can compose with stdlib `.and()` / `.or()` / `.negate()`. No new DSL to learn.

## 🛣️ Where this lands on the roadmap

V4 (#160) is currently in `status:backlog` — Phase 0–6, ~18–22 weeks. Your proposal becomes **Phase 7: Navigational API on top of ScreenContext** — small phase, ~2 weeks, lands once the detection pipeline is stable. A sub-issue under #160 will be opened to track it explicitly.

## 🌏 The Chinese double-touch this week

It's not lost on me that this comment arrives **in Chinese** the same week we shipped Unicode-aware `type()` for Chinese / Japanese / Korean input (#232 — opened by another Chinese contributor, @blackball, ten days earlier). The international OculiX community is finally being served on **all three sides** of the same coin:

- **Reading** the UI in any language → OCR via bundled `chi_sim` / `jpn` / `kor` (V4 Phase 4 dataset, on top of the v3.0.3-rc5 #233 release)
- **Typing** in any language → `type("你好")` works since rc5 #232
- **Navigating** the UI semantically → the `UITree` you just sketched, V4

Three Chinese contributors (@blackball reading + typing in #232, you navigating in #160), three pieces of the puzzle, in the same release cycle. This is the open-source dynamic at its best — community ahead of the maintainer.

---

```
                       ⠀⠀⠀⠀⠀⠀⠀⠀⣀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
                       ⠀⠀⠀⠀⠀⠀⣰⣿⣿⣿⣿⣦⣀⣀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
                       ⠀⠀⠀⠀⠀⠀⢿⣿⠟⠋⠉⠀⠀⠀⠀⠉⠑⠢⣄⡀⠀⠀⠀⠀⠀
                       ⠀⠀⠀⠀⠀⢠⠞⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠙⢿⣿⣿⣦⡀
                       ⠀⣀⠀⠀⢀⡏⠀⢀⣴⣶⣶⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⢻⣿⣿⠇
                       ⣾⣿⣿⣦⣼⡀⠀⢺⣿⣿⡿⠃⠀⠀⠀⠀⣠⣤⣄⠀⠀⠈⡿⠋⠀
                       ⢿⣿⣿⣿⣿⣇⠀⠤⠌⠁⠀⡀⢲⡶⠄⢸⣏⣿⣿⠀⠀⠀⡇⠀⠀
                       ⠈⢿⣿⣿⣿⣿⣷⣄⡀⠀⠀⠈⠉⠓⠂⠀⠙⠛⠛⠠⠀⡸⠁⠀⠀
                       ⠀⠀⠻⣿⣿⣿⣿⣿⣿⣷⣦⣄⣀⠀⠀⠀⠀⠑⠀⣠⠞⠁⠀⠀⠀
                       ⠀⠀⠀⢸⡏⠉⠛⠛⠛⠿⠿⣿⣿⣿⣿⣿⣿⣿⣿⡄⠀⠀⠀⠀⠀
                       ⠀⠀⠀⠸⠀⠀⠀⠀⠀⠀⠀⠀⠈⠉⠛⢿⣿⣿⣿⣿⡄⠀⠀⠀⠀
                       ⠀⠀⠀⢷⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢻⣿⣿⣿⣿⡀⠀⠀⠀
                       ⠀⠀⠀⢸⣆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣿⣿⣿⣿⡇⠀⠀⠀
                       ⠀⠀⠀⢸⣿⣦⣀⠀⠀⠀⠀⠀⠀⠀⠀⢀⣼⡟⠻⠿⠟⠀⠀⠀⠀
                       ⠀⠀⠀⠀⣿⣿⣿⣿⣶⠶⠤⠤⢤⣶⣾⣿⣿⡇⠀⠀⠀⠀⠀⠀⠀
                       ⠀⠀⠀⠀⠹⣿⣿⣿⠏⠀⠀⠀⠈⢿⣿⣿⡿⠁⠀⠀⠀⠀⠀⠀⠀
                       ⠀⠀⠀⠀⠀⠈⠉⠉⠀⠀⠀⠀⠀⠀⠉⠉⠀⠀⠀⠀⠀⠀⠀⠀⠀

                  ✓ The panda is watching. The UITree is coming.
                            The community is ahead of us.
```

---
